### ISSUE TITLE
DPL-437-7 Billing report: Refactor billing query

### ISSUE THREAD

**Issue by harrietc52** [sanger/unified_warehouse#414]
**User story**
As a developer, I would like to refactor the billing query. Create a temporary table for 2 of the nested queries:
1. Get the first run
2. Get how many samples and controls are present in a lane

**Who are the primary contacts for this story**
@harrietc52 

**Who is the nominated tester for UAT**
e.g. John S (don't include surnames in public repos)

**Acceptance criteria**
To be considered successful the solution must allow:
- [ ] Create a tmp table to subquery 1
- [ ] Create a tmp table to subquery 2
- [ ] Use these queries in the main Billing query
- [ ] Add comments
- [ ] Syntax format is correct
- [ ] Update billing query in repo https://github.com/sanger/unified_warehouse/issues/410
- [ ] Update stored procedure

**Dependencies**
This story is blocked by the following dependencies:
- https://github.com/sanger/unified_warehouse/issues/409

References
This story has a non-blocking relationship with:
- This is a spin-out story from https://github.com/sanger/unified_warehouse/issues/386

**Additional context**

**Comment by harrietc52** [comment]
**Common Table Expressions**

> I am not sure if Katy asked some references about the with statement (CTE, Common Table Expressions).  Sorry I don't have good links but Google search shows so many examples in different DBs.
I learned it by doing it but I looked at the Postgres docs time-to-time. https://www.postgresql.org/docs/current/queries-with.html . I did not use parametrised or recursive versions much.
In simple use, it is like creating views on the fly, to refer later in the main query like joining them together. This is especially useful if there are multiple confusing group-bys.
Another use was like refining steps, define A, define B using A, ..., and get the final result in the last query.
My main concern was trying to keep the intermediate result sets small to avoid big scans later. I used Explain statement to  get an idea about the cost of the original query and the new ones. Roughly, trying to favour index use, less scans and less number of intermediate records. At one point, there was a need to extract some computations into C extensions to be able to do better math and control.

**Comment by harrietc52** [comment]
@TWJW-SANGER

**Comment by TWJW-SANGER** [comment]
In the current MLWH setup creating temp tables requires write access, which we probably don't want to enforce for a stored Proc.
CTE is not available in MySQL 5.7.

So I think this is a nice to have, that can't actually be done as it stands. I am happy to close and fold into any further billing enhancements as and when possible.
